rust/src/linearview.rs

```diff
@@ -0,0 +1,515 @@
+    }
+
+    pub fn single_function_hlil(function: &Function, settings: &DisassemblySettings) -> Ref<Self> {
+        unsafe {
+            let handle = binaryninjacore_sys::BNCreateLinearViewSingleFunctionHighLevelIL(
+                function.handle,
+                settings.handle,
+            );
+
+            Self::from_raw(handle)
+        }
+    }
+
+    pub fn single_function_hlil_ssa(
+        function: &Function,
+        settings: &DisassemblySettings,
+    ) -> Ref<Self> {
+        unsafe {
+            let handle = binaryninjacore_sys::BNCreateLinearViewSingleFunctionHighLevelILSSAForm(
+                function.handle,
+                settings.handle,
+            );
+
+            Self::from_raw(handle)
+        }
+    }
+
+    pub fn single_function_language_representation(
+        function: &Function,
+        settings: &DisassemblySettings,
+    ) -> Ref<Self> {
+        unsafe {
+            let handle =
+                binaryninjacore_sys::BNCreateLinearViewSingleFunctionLanguageRepresentation(
+                    function.handle,
+                    settings.handle,
+                );
+
+            Self::from_raw(handle)
+        }
+    }
+}
+
+unsafe impl RefCountable for LinearViewObject {
+    unsafe fn inc_ref(handle: &Self) -> Ref<Self> {
+        Ref::new(Self {
+            handle: BNNewLinearViewObjectReference(handle.handle),
+        })
+    }
+
+    unsafe fn dec_ref(handle: &Self) {
+        BNFreeLinearViewObject(handle.handle);
+    }
+}
+
+impl ToOwned for LinearViewObject {
+    type Owned = Ref<Self>;
+
+    fn to_owned(&self) -> Self::Owned {
+        unsafe { RefCountable::inc_ref(self) }
+    }
+}
+
+unsafe impl Send for LinearViewObject {}
+unsafe impl Sync for LinearViewObject {}
+
+pub struct LinearViewObjectIdentifier {
+    handle: *mut BNLinearViewObjectIdentifier,
+}
+
+impl LinearViewObjectIdentifier {
+    #[allow(unused)]
+    pub(crate) unsafe fn from_raw(handle: *mut BNLinearViewObjectIdentifier) -> Self {
+        debug_assert!(!handle.is_null());
+        Self { handle }
+    }
+}
+
+impl Drop for LinearViewObjectIdentifier {
+    fn drop(&mut self) {
+        unsafe { BNFreeLinearViewObjectIdentifier(self.handle) }
+    }
+}
+
+#[derive(Eq)]
+pub struct LinearViewCursor {
+    pub(crate) handle: *mut binaryninjacore_sys::BNLinearViewCursor,
+}
+
+impl LinearViewCursor {
+    pub(crate) unsafe fn from_raw(handle: *mut BNLinearViewCursor) -> Ref<Self> {
+        debug_assert!(!handle.is_null());
+
+        Ref::new(Self { handle })
+    }
+
+    pub fn new(root: &LinearViewObject) -> Ref<Self> {
+        unsafe {
+            let handle = BNCreateLinearViewCursor(root.handle);
+            Self::from_raw(handle)
+        }
+    }
+
+    // FIXME: can we implement clone without shadowing ToOwned?
+    pub fn duplicate(&self) -> Ref<Self> {
+        unsafe {
+            let handle = BNDuplicateLinearViewCursor(self.handle);
+            Self::from_raw(handle)
+        }
+    }
```

Comment:
I don't _think_ so, if I'm understanding you completely.  You get a completely new object in the core, so you have to own the object outside of it as well.  That said, this should probably be moved to a `impl Clone`

Reply:
This is what I originally intended, but that conflicted with some blanket impl from the refcounting.

Reply:
FWIW, here's what happens when I try to `impl Clone`:
```
error[E0119]: conflicting implementations of trait `std::borrow::ToOwned` for type `linearview::LinearViewCursor`
   --> /Users/user/Code/binaryninja-api/rust/src/linearview.rs:335:1
    |
335 | impl ToOwned for LinearViewCursor {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: conflicting implementation in crate `alloc`:
            - impl<T> ToOwned for T
              where T: Clone;

For more information about this error, try `rustc --explain E0119`.
error: could not compile `binaryninja` due to previous error
```